% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/plots.R
\name{gg.manhattan}
\alias{gg.manhattan}
\title{Manhattan plot function for gwas (stolen from pcgoddard)}
\usage{
gg.manhattan(
  df,
  hlight,
  col = "default",
  title = "Manhattan",
  SNP = "SNP",
  CHR = "CHR",
  BP = "BP",
  P = "P",
  sig = 5e-08,
  sugg = 1e-05,
  lab = FALSE,
  colour = TRUE,
  remove_chr_labs = NULL
)
}
\arguments{
\item{df}{data.frame containing your results}

\item{hlight}{a vector containing any SNPs that need highlighting}

\item{col}{colours for the plot (default = RColorBrewer::brewer.pal(9, "Greys")[c(5,9)])}

\item{title}{title of the plot}

\item{SNP}{name of the column corresponding to SNPs}

\item{CHR}{name of the column corresponding to chromosome number}

\item{BP}{name of the column corresponding to base pair}

\item{P}{name of the column corresponding to p value}

\item{sig}{the "significance threshold"}

\item{sugg}{the "suggestive significance threshold"}

\item{lab}{whether or not to add labels to the sites on the plot}

\item{colour}{coloured plot?}

\item{remove_chr_labs}{chromosome labels to remove from the x-axis. default = NULL}
}
\value{
manhattan plot
}
\description{
Generates a Manhattan plot using gwas data
}
